File: backend/app/models/debate.py
from pydantic import BaseModel
from typing import Optional
from enum import Enum

class AgentKey(str, Enum):
    REALIST  = "realist"
    LOGICIAN = "logician"
    MIRROR   = "mirror"
    STEELMAN = "steelman"

class AgentResponse(BaseModel):
    agent: AgentKey
    content: str
    round: int

class RoundHistory(BaseModel):
    round: int
    responses: list[AgentResponse] = []
    rebuttal: Optional[str] = None

class DebateSession(BaseModel):
    session_id: str
    idea: str
    current_round: int = 0
    max_rounds: int = 3
    rounds: list[RoundHistory] = []
    # Flat list of all agent responses for report generation
    all_responses: dict[str, list[str]] = {}
    status: str = "active"  # active | complete

class StartDebateRequest(BaseModel):
    idea: str

class RebuttalRequest(BaseModel):
    rebuttal: str

class ResilienceReport(BaseModel):
    realist_score: float
    logician_score: float
    mirror_score: float
    overall_score: float
    open_weaknesses: str
    defended_well: str
    original_pitch: str
    refined_pitch: str

File: frontend/src/store/debate.ts
import { create } from "zustand";

export type AgentKey = "realist" | "logician" | "mirror" | "steelman";

export interface AgentResponse {
  agent: AgentKey;
  content: string;
  round: number;
}

export interface Round {
  responses: AgentResponse[];
  rebuttal?: string;
}

interface DebateStore {
  // State
  sessionId: string | null;
  idea: string;
  rounds: Round[];
  currentRound: number;
  status: "idle" | "loading" | "awaiting_rebuttal" | "generating_report" | "complete";
  report: Record<string, unknown> | null;

  // Actions
  setIdea: (idea: string) => void;
  setSessionId: (id: string) => void;
  addAgentResponse: (r: AgentResponse) => void;
  setRebuttal: (roundIndex: number, text: string) => void;
  incrementRound: () => void;
  setStatus: (s: DebateStore["status"]) => void;
  setReport: (r: Record<string, unknown>) => void;
  reset: () => void;
}

const initialState = {
  sessionId: null, idea: "", rounds: [],
  currentRound: 0, status: "idle" as const, report: null,
};

export const useDebate = create<DebateStore>((set) => ({
  ...initialState,

  setIdea: (idea) => set({ idea }),
  setSessionId: (sessionId) => set({ sessionId }),

  addAgentResponse: (r) =>
    set((state) => {
      const rounds = [...state.rounds];
      const idx = r.round - 1;
      if (!rounds[idx]) rounds[idx] = { responses: [] };
      rounds[idx] = {
        ...rounds[idx],
        responses: [...rounds[idx].responses, r],
      };
      return { rounds };
    }),

  setRebuttal: (roundIndex, text) =>
    set((state) => {
      const rounds = [...state.rounds];
      if (rounds[roundIndex]) rounds[roundIndex] = { ...rounds[roundIndex], rebuttal: text };
      return { rounds };
    }),

  incrementRound: () =>
    set((state) => ({ currentRound: state.currentRound + 1 })),

  setStatus: (status) => set({ status }),
  setReport: (report) => set({ report, status: "complete" }),
  reset: () => set(initialState),
}))